clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -202,6 +210,40 @@ public CompletableFuture<Void> maybeAutoCommitAllConsumed() {
         return maybeAutoCommit(subscriptions.allConsumed());
     }
 
+    /**
+     * The consumer needs to send an auto commit during the shutdown if autocommit is enabled.
+     */
+    Optional<NetworkClientDelegate.UnsentRequest> maybeCreateAutoCommitRequest() {
+        if (!autoCommitState.isPresent()) {
+            return Optional.empty();
+        }
+
+        OffsetCommitRequestState request = pendingRequests.createOffsetCommitRequest(subscriptions.allConsumed(), jitter);
+        request.future.whenComplete(autoCommitCallback(subscriptions.allConsumed()));
+        return Optional.of(request.toUnsentRequest());
+    }
+
+    private CompletableFuture<Void> sendAutoCommit(final Map<TopicPartition, OffsetAndMetadata> allConsumedOffsets) {
+        log.debug("Enqueuing autocommit offsets: {}", allConsumedOffsets);
+        return addOffsetCommitRequest(allConsumedOffsets).whenComplete(autoCommitCallback(allConsumedOffsets));
+    }
+
+    private BiConsumer<? super Void, ? super Throwable> autoCommitCallback(final Map<TopicPartition, OffsetAndMetadata> allConsumedOffsets) {
+        return (response, throwable) -> {
+            autoCommitState.ifPresent(autoCommitState -> autoCommitState.setInflightCommitStatus(false));
```

Comment:
It is checked by `canSendAutocommit` right?

Reply:
Good point, missed that. Do we not want to check it when creating the auto commit request?

Reply:
Perhaps you meant by this: `maybeCreateAutoCommitRequest`

Debatable. Since we aren't sure what offset the previous autocommit was sending (if there is an inflight one) during the shutdown, we want to send an autocommit regardless.

In fact, now makes me think, we might not need to check if there is an inflight autocommit at all, as we always want to make sure committing an offset on the clock.  I'll file a jira for investigation.  Thanks!

Reply:
Here's the ticket:https://issues.apache.org/jira/browse/KAFKA-15872

Reply:
Yeah, makes sense. We are just going to end up in a weird state here because we may have multiple autocommits in-flight, but when the first one returns we will set `inFlightCommitStatus` to false, although there are more in-flight. I can't point to anything that will break, but it seems a little off.